openedx/openedx-platform
# (WIP) Fix TNL-3429.

**JIRA tickets**: [TNL-3429](https://openedx.atlassian.net/browse/TNL-3429)

**Discussion**: 
This PR replaces https://github.com/edx/edx-platform/pull/10756.  It builds on work done by @robrap in https://github.com/edx/edx-platform/pull/10962, for [TNL-3425](https://openedx.atlassian.net/browse/TNL-3425).

These changes aim to follow https://openedx.atlassian.net/wiki/display/TNL/XSS+Protection .

The change to cms/templates/container.html replaces calls to display_name_with_default_escaped, which is deprecated and does not properly perform HTML-escaping, with display_name_with_default, which makes use of the |h filter for escaping.

The changes to these three files selectively apply escaping, using either the deprecated display_name_with_default_escaped or the preferred display_name_with_default, depending on the context, so that only the course outline and unit pages get the non-escaped data:
- cms/djangoapps/contentstore/views/component.py
- cms/djangoapps/contentstore/views/course.py
- cms/djangoapps/contentstore/views/item.py

These two templates are changed to HTML-escape the display name fields:
- cms/templates/js/course-outline.underscore
- cms/templates/js/unit-outline.underscore

**Dependencies**: None

**Sandbox**: [LMS](http://pr11588.sandbox.opencraft.com/) [Studio](http://studio.pr11588.sandbox.opencraft.com/)

**Testing instructions**:
_Note: in the screenshots, the fields of interest are highlighted in purple._
1. Start up Studio as the edxapp user: "paver devstack studio".
2. Open a browser to Studio: http://127.0.0.1:8001. The home page loads.
3. Click the "Sign in" button. The sign in page loads.
4. Sign in as as the "staff@example.com" user.
5. Click on "edX Demonstration Course". This opens the course outline page.
6. Scroll to the bottom of the page.
7. Click "New Section" and create a section named `§: Intro to <HTML> & "JS"`
8. Click "New Subsection" and create a subsection named `§§: Intro to <HTML> & "JS"`. Both names are rendered correctly on the course outline page.
   ![outline-page-loaded-before](https://cloud.githubusercontent.com/assets/6389399/13157632/ab99d284-d691-11e5-99b4-acfdece15435.png)
9. If necessary, click on the triangle alongside the subsection name to expand the field.
10. Click "New Unit". The unit page loads.
11. Without the fix, this reproduces bug TNL-3429:
    ![unit-page-loaded-before](https://cloud.githubusercontent.com/assets/6389399/13157716/2a946c66-d692-11e5-9b56-0f3df0e138de.png)
12. With the fix, the text is rendered correctly:
    ![unit-page-loaded-after](https://cloud.githubusercontent.com/assets/6389399/13157725/33bec066-d692-11e5-8a60-4904a0d679c1.png)
13. Return to the course outline page.  Click to edit the section name.  Without the fix, the text is rendered in the text-field incorrectly (the < and > characters are HTML-escaped, but not the & and " characters, which appear as expected; the partial escaping is obvious with this particular test-data):
    ![outline-page-edited-before](https://cloud.githubusercontent.com/assets/6389399/13157651/c20af1c4-d691-11e5-946f-7d8544e7a9e9.png)

With these changes, the text is rendered as correctly:
![outline-page-edit-after](https://cloud.githubusercontent.com/assets/6389399/13157874/e6b7e4d6-d692-11e5-8012-884157d17550.png)

This editing bug occurs on the unit page too, and is also fixed by this change.  I am not aware of any JIRA ticket tracking this issue.

**Author notes and concerns**:
This PR changes the  way that display_name text is sent via JSON. Previously, the < and > characters would be escaped to `&lt;` and `&gt;`. No other HTML-sensitive characters were handled.  Now, these characters are sent as-is.  For the course outline page and the unit page, the escaping to HTML is performed on the browser by the underscore templates (properly, not just for angle-brackets).

**Reviewers**
- [ ] @smarnach, TBD
